docs/eBPFWeakReferences.md

```diff
@@ -0,0 +1,82 @@
+
+The weak-ref implementation extends the base "ebpf object" management layer and is thus available for all primary object types.  The weak-ref management API are prototyped in the ```epbf_object.h``` header and provides the following calls:
+
+
+```ebpf_weak_reference_t* ebpf_object_weak_reference_get_reference(ebpf_core_object_t* object);```
+<br>
+The caller typically begins by calling this API to get a weak-ref pointer to the primary object they are interested in. This call increments the ref count for the weak-ref but does not modify the primary object's ref-count.  The caller now has a 'token' that they can now use to (conditionally) get the primary object pointer, either right away or at some later point in time.
+<br><br>
+
+```ebpf_core_object_t* ebpf_object_weak_reference_get_object_reference(ebpf_weak_reference_t* weak_reference);```
+<br>
+The caller uses this call to try and get the primary object pointer.  The 'try' is deliberate in that this call is not guaranteed to return a valid pointer to the referred-to object as the referred-to object might have been destroyed (or is in the process of being destroyed).  In such cases, this call will return a NULL.  If the object is still around, this call increments the ref-count of the referred-to object and returns a valid object pointer to the caller. The ref-count of the weak-ref object is not modified in this entire process. In case of a valid pointer return, the caller is responsible for releasing the reference on the returned object.
+<br><br>
+
+```void ebpf_object_weak_reference_acquire_reference(ebpf_weak_reference_t* weak_reference);```
+<br>
+The caller uses this call to increment the ref-count of the weak-ref object, to ensure the weak-ref's validity and/or longevity in multiple reference scenarios.
+<br><br>
+
+```void ebpf_object_weak_reference_release_reference(ebpf_weak_reference_t* weak_reference);```
+<br>
+The caller uses this call to decrement the ref-count of the weak-ref object, when the weak-ref object is no longer needed. When the ref-count of the weak-ref goes to zero, it is destroyed.  Note that the weak-ref starts life with a ref-count of 1 when it is created and associated with the referred-to object (in the context of it's referred-to object's creation) and hence is guaranteed to stay alive until after the primary object is destroyed.
+
+
+**Conclusion.**
+
+The weak-ref design provides a simple and maintainable solution to the circular reference problem.  With this approach, the object maps (Prog-Array, Array-Map-of-Maps, Hash-Map-of-Maps) do not need to take a strong reference on the contained objects anymore.  Instead, they now acquire their weak-ref equivalents and store them along side their stored object ids (the per-entry storage space has been suitably expanded to accomodate the weak-ref pointer). When needed, they can now try to get pointers to the referred-to primary objects with the explicit understanding that they might not succeed in doing so.
```

Comment:
```suggestion
The weak-ref design provides a simple and maintainable solution to the circular reference problem.  With this approach, the object maps (Prog-Array, Array-Map-of-Maps, Hash-Map-of-Maps) do not need to take a strong reference on the contained objects anymore.  Instead, they now acquire their weak-ref equivalents and store them alongside their stored object ids (the per-entry storage space has been suitably expanded to accommodate the weak-ref pointer). When needed, they can now try to get pointers to the referred-to primary objects with the explicit understanding that they might not succeed in doing so.
```

The question I would ask is if this design would require a map or program to need pinning on Windows where it would not need pinning on Linux due to a strong reference being taken.  If not, great, but if so then that would harm cross-platform programs.

Reply:
Investigating